components/raftstore/src/store/fsm/apply.rs

```diff
@@ -2772,160 +2617,122 @@ where
+        let (region_id, peer_id) = (self.region_id(), self.id());
         let propose_num = props_drainer.len();
-        if self.delegate.stopped {
+        if self.stopped {
             for p in props_drainer {
                 let cmd = PendingCmd::<EK::Snapshot>::new(p.index, p.term, p.cb);
-                notify_stale_command(region_id, peer_id, self.delegate.term, cmd);
+                notify_stale_command(region_id, peer_id, self.term, cmd);
             }
             return;
         }
         for p in props_drainer {
             let cmd = PendingCmd::new(p.index, p.term, p.cb);
             if p.is_conf_change {
-                if let Some(cmd) = self.delegate.pending_cmds.take_conf_change() {
+                if let Some(cmd) = self.pending_cmds.take_conf_change() {
                     // if it loses leadership before conf change is replicated, there may be
                     // a stale pending conf change before next conf change is applied. If it
                     // becomes leader again with the stale pending conf change, will enter
                     // this block, so we notify leadership may have been changed.
-                    notify_stale_command(region_id, peer_id, self.delegate.term, cmd);
+                    notify_stale_command(region_id, peer_id, self.term, cmd);
                 }
-                self.delegate.pending_cmds.set_conf_change(cmd);
+                self.pending_cmds.set_conf_change(cmd);
             } else {
-                self.delegate.pending_cmds.append_normal(cmd);
+                self.pending_cmds.append_normal(cmd);
             }
         }
         // TODO: observe it in batch.
         APPLY_PROPOSAL.observe(propose_num as f64);
     }
 
     fn destroy<W: WriteBatch<EK>>(&mut self, ctx: &mut ApplyContext<EK, W>) {
-        let region_id = self.delegate.region_id();
+        let region_id = self.region_id();
         if ctx.apply_res.iter().any(|res| res.region_id == region_id) {
             // Flush before destroying to avoid reordering messages.
             ctx.flush();
         }
-        fail_point!(
-            "before_peer_destroy_1003",
-            self.delegate.id() == 1003,
-            |_| {}
-        );
+        fail_point!("before_peer_destroy_1003", self.id() == 1003, |_| {});
         info!(
             "remove delegate from apply delegates";
-            "region_id" => self.delegate.region_id(),
-            "peer_id" => self.delegate.id(),
+            "region_id" => self.region_id(),
+            "peer_id" => self.id(),
         );
-        self.delegate.destroy(ctx);
+        self.stopped = true;
+        for cmd in self.pending_cmds.normals.drain(..) {
+            notify_region_removed(self.region.get_id(), self.id, cmd);
+        }
+        if let Some(cmd) = self.pending_cmds.conf_change.take() {
+            notify_region_removed(self.region.get_id(), self.id, cmd);
+        }
```

Comment:
Missing `apply_ctx.router.close(self.region_id())` from original `self.delegate.destroy(ctx)` implement.

Reply:
No, we do not need `router` any more. Because there is no `ApplyRouter` which include routers of all regions.